src/usbg.c

```diff
@@ -2197,6 +2447,94 @@ int usbg_get_binding_name_s(usbg_binding *b, char *buf, int len)
 	return snprintf(buf, len, "%s", b->name);
 }
 
+static int usbg_create_os_desc_link(usbg_gadget *g, usbg_config *c)
+{
+	char bpath[USBG_MAX_PATH_LENGTH];
+	char cpath[USBG_MAX_PATH_LENGTH];
+	int nmb;
+	int ret;
+
+	nmb = snprintf(cpath, sizeof(cpath), "%s/%s", c->path, c->name);
+	if (nmb >= sizeof(cpath)) {
+		ret = USBG_ERROR_PATH_TOO_LONG;
+		goto out;
+	}
+
+	nmb = snprintf(bpath, sizeof(bpath), "%s/%s/%s/%s", g->path, g->name,
+			OS_DESC_DIR, c->name);
+	if (nmb >= sizeof(bpath)) {
+		ret = USBG_ERROR_PATH_TOO_LONG;
+		goto out;
+	}
+
+	ret = symlink(cpath, bpath);
+	if (ret != 0) {
+		ret = usbg_translate_error(errno);
+		goto out;
+	}
+
+	g->os_desc_binding = c;
+
+	return USBG_SUCCESS;
+out:
+	return ret;
+}
+
+static int usbg_rm_os_desc_link(usbg_gadget *g)
+{
+	char bpath[USBG_MAX_PATH_LENGTH];
+	struct dirent **dent;
+	int end, n, i;
+	int ret;
+
+	end = snprintf(bpath, sizeof(bpath), "%s/%s/%s", g->path, g->name,
+		       OS_DESC_DIR);
+	if (end >= sizeof(bpath)) {
+		ret = USBG_ERROR_PATH_TOO_LONG;
+		goto out;
+	}
+
+	n = scandir(bpath, &dent, bindings_select, alphasort);
+	if (n < 0) {
+		ret = usbg_translate_error(errno);
+		goto out;
+	}
+
+	if (n < 1) {
+		ret = USBG_ERROR_NOT_FOUND;
+		goto out;
```

Comment:
Shouldn't we also free(dent) in this case? The example in https://linux.die.net/man/3/scandir seems to do it.

Reply:
Done. Thanks